crates/ethrpc/src/buffered.rs

```diff
@@ -84,23 +85,50 @@ where
         tokio::task::spawn(batched_for_each(config, calls, move |batch| {
             let inner = inner.clone();
             async move {
-                let (mut requests, mut senders): (Vec<_>, Vec<_>) = batch
-                    .into_iter()
-                    .filter(|(_, _, sender)| !sender.is_canceled())
-                    .map(|(id, request, sender)| ((id, request), sender))
-                    .unzip();
+                let (mut requests, mut trace_ids, mut senders): (Vec<_>, Vec<_>, Vec<_>) =
+                    itertools::multiunzip(
+                        batch
+                            .into_iter()
+                            .filter(|(_, _, _, sender)| !sender.is_canceled())
+                            .map(|(id, request, trace_id, sender)| {
+                                ((id, request), trace_id, sender)
+                            }),
+                    );
                 match requests.len() {
                     0 => (),
                     1 => {
-                        let ((id, request), sender) = (requests.remove(0), senders.remove(0));
-                        let result = inner.send(id, request).await;
+                        let ((id, request), trace_id, sender) =
+                            (requests.remove(0), trace_ids.remove(0), senders.remove(0));
+                        let result = match trace_id {
+                            Some(trace_id) => {
+                                observe::request_id::set_task_local_storage(
+                                    trace_id,
+                                    inner.send(id, request),
+                                )
+                                .await
+                            }
+                            None => inner.send(id, request).await,
+                        };
                         let _ = sender.send(result);
                     }
                     n => {
-                        let results = inner
-                            .send_batch(requests)
-                            .await
-                            .unwrap_or_else(|err| vec![Err(err); n]);
+                        let request_metadata = requests
+                            .iter()
+                            .zip(trace_ids)
+                            .filter_map(|((_, call), trace_id)| match (call, trace_id) {
+                                (Call::MethodCall(call), Some(trace_id)) => {
+                                    Some(format!("{}:{}", trace_id, call.method))
+                                }
+                                _ => None,
+                            })
+                            .collect_vec()
+                            .join(",");
+                        let results = observe::request_id::set_task_local_storage(
```

Comment:
I believe just overwriting the existing value can lead to mismatched requests metadata when the consumer is slower than the producer. Imagine this:
1. buffered transport writes requests and metadata into channel
2. consumer channel is busy handling prepared requests
3. more requests come in and the buffered transport overwrites task local storage with latest metadata
4. consumer channel is done and picks requests from `1` but metadata from `3`

I think the task local storage also needs to use a queue to avoid getting these mismatches.

Reply:
But [batches.for_each_concurrent(concurrency_limit, work)](https://github.com/cowprotocol/services/blob/0b7887ed7011f400b387f4514c1dd9522050a40e/crates/ethrpc/src/buffered.rs#L258) executes each batch in a separate task, which sets its own storage, doesn't it?

Reply:
I don't understand the concern (but maybe my understanding of the concurrency is wrong). 
Task local storage is set for the scope of the inner future which we await. How would it get overridden by some other task? If this was possible, wouldn't this mean our top level tracing spans (which also use task local storage) could get overridden by concurrent API requests that execute on the same thread?

Reply:
If I got it correctly, the main concern here is how `batched_for_each` is implemented. We run this function inside a single tokio task context, where internally each batch is executed concurrently in a separate future. Since the tokio task storage is now shared across multiple concurrent futures, a race condition occurs.

Reply:
Nevermind I misunderstood how the ids get moved into the task that actually sends the request.
For some reason I thought the background task would read the task local storage of the task that enqueues the batches. But what actually happens is:
1. task created for user request wants to issues an RPC request
2. `buffered` reads local storage of the current task (no race condition) and adds it to the context
3. context gets sent into background task via a channel (so task local storage is now irrelevant also this is effectively a queue)
4. background tasks reads RPC requests from the channel which provides the necessary context for each call
5. background task batches multiple RPC requests with all their IDs

So I think https://github.com/cowprotocol/services/pull/2750/commits/de7d920b3f5caabdb251c3d3ebde4b0e362d7fdb can be reverted after all. Sorry for the false alarm. 😅

Reply:
But what about executing multiple futures(for each batch) within a single tokio task context? These futures are executed concurrently(not sequentially), meaning they can all make progress within overlapping time periods.

Reply:
I think by using `observe::request_id::set_task_local_storage` you already guarantee that whenever the future gets polled it first sets the task local storage to the correct value before the inner future can make progress which might produce some logs.

Reply:
A single Tokio task is spawned, and then N futures with some concurrency level start executing. All the futures share the same tokio task context. No additional tokio tasks are spawned there.

First operation of each future is to override that shared context. When a future in a Tokio task hits an `.await`, it yields control back to the scheduler, where the scheduler might switch to another future, which overrides the context before the first future reads it.

We have the first `.await` call [even before](https://github.com/cowprotocol/services/blob/0b7887ed7011f400b387f4514c1dd9522050a40e/crates/ethrpc/src/http.rs#L144) accessing the Tokio task context. 

Still not sure what exactly guarantees that is not possible to happen.

Reply:
```
background_worker() executes:
    observe::request_id::set_task_local_storage(
        trace_id,
        inner.send(id, request),
    )

Inner::send() executes:
    async move { this.execute_call(id, request).await }
        .in_current_span()
        .boxed()

BufferedTransport::execute_call() executes:
    let response = self.queue_call(id, request);

BufferedTransport::queue_call() executes:
        let trace_id = observe::request_id::get_task_local_storage();
        let context = (id, request, trace_id, sender);
        self.calls
            .unbounded_send(context)
            .expect("worker task unexpectedly dropped");
```


All the logic executed for this batch of calls happens in the same future because we don't spawn any new tasks internally. And all of that gets wrapped in `set_task_local_storage()`

> We have the first .await call even before accessing the Tokio task context.

If I'm reading this correctly you are worried that we might yield the future before we event get to reading its local storage and when we switch back to that future some other future might have already overwritten the local storage in the meantime and now we read the wrong value. Is that your concern?

This will not happen because `set_task_local_storage()` does not just set the task local storage when the future gets created but every time the scheduler switches to it to make progress on it (`Future::poll()`).
We could work on 100 different futures in the mean time which all overwrite the local storage but when we get back to the one we are looking at the first thing we do is to set the local storage to the value we expect.

Reply:
Omg, thanks a lot, now I got it. Also, created a local test, and it proves that the `set_task_local_storage()` is called on each `Future::poll()`. e.g. for 1000 elements it was called >3000 times(instead of 1000). Rust's runtime model is impressive.